ingestion/src/main/java/feast/ingestion/transform/metrics/WriteFeatureValueMetricsDoFn.java

```diff
@@ -0,0 +1,312 @@
+    }
+  }
+
+  // Update stats and values array for the feature represented by the field.
+  // If the field contains non-numerical or non-boolean value, the stats and values array
+  // won't get updated because we are only concerned with numerical value in metrics data.
+  // For boolean value, true and false are treated as numerical value of 1 of 0 respectively.
+  private void updateStats(
+      Map<String, DoubleSummaryStatistics> featureNameToStats,
+      Map<String, List<Double>> featureNameToValues,
+      Field field) {
+    if (featureNameToStats == null || featureNameToValues == null || field == null) {
+      return;
+    }
+
+    String featureName = field.getName();
+    if (!featureNameToStats.containsKey(featureName)) {
+      featureNameToStats.put(featureName, new DoubleSummaryStatistics());
+    }
+    if (!featureNameToValues.containsKey(featureName)) {
+      featureNameToValues.put(featureName, new ArrayList<>());
+    }
+
+    Value value = field.getValue();
+    DoubleSummaryStatistics stats = featureNameToStats.get(featureName);
+    List<Double> values = featureNameToValues.get(featureName);
+
+    switch (value.getValCase()) {
+      case INT32_VAL:
+        stats.accept(value.getInt32Val());
+        values.add(((double) value.getInt32Val()));
+        break;
+      case INT64_VAL:
+        stats.accept(value.getInt64Val());
+        values.add((double) value.getInt64Val());
+        break;
+      case DOUBLE_VAL:
+        stats.accept(value.getDoubleVal());
+        values.add(value.getDoubleVal());
+        break;
+      case FLOAT_VAL:
+        stats.accept(value.getFloatVal());
+        values.add((double) value.getFloatVal());
+        break;
+      case BOOL_VAL:
+        stats.accept(value.getBoolVal() ? 1 : 0);
+        values.add(value.getBoolVal() ? 1d : 0d);
+        break;
+      case INT32_LIST_VAL:
+        for (Integer val : value.getInt32ListVal().getValList()) {
+          stats.accept(val);
+          values.add(((double) val));
+        }
+        break;
+      case INT64_LIST_VAL:
+        for (Long val : value.getInt64ListVal().getValList()) {
+          stats.accept(val);
+          values.add(((double) val));
+        }
+        break;
+      case DOUBLE_LIST_VAL:
+        for (Double val : value.getDoubleListVal().getValList()) {
+          stats.accept(val);
+          values.add(val);
+        }
+        break;
+      case FLOAT_LIST_VAL:
+        for (Float val : value.getFloatListVal().getValList()) {
+          stats.accept(val);
+          values.add(((double) val));
+        }
+        break;
+      case BOOL_LIST_VAL:
+        for (Boolean val : value.getBoolListVal().getValList()) {
+          stats.accept(val ? 1 : 0);
+          values.add(val ? 1d : 0d);
+        }
+        break;
+      case BYTES_VAL:
+      case BYTES_LIST_VAL:
+      case STRING_VAL:
+      case STRING_LIST_VAL:
+      case VAL_NOT_SET:
```

Comment:
What happens if the value is not set? Are we counting those?

Reply:
It is already counted here (I think it's ok not pre-aggregate that part for now)
https://github.com/gojek/feast/blob/aec7979f2986e56d7be525cc6b27a1eeb786d501/ingestion/src/main/java/feast/ingestion/transform/metrics/WriteRowMetricsDoFn.java#L121